modules/backend/models/User.php

```diff
@@ -103,17 +103,12 @@ public function getAvatarThumb($size = 25, $options = null)
             $options = [];
         }
 
-        // Default is "mm" (Mystery man)
-        $default = array_get($options, 'default', 'mm');
-
         if ($this->avatar) {
             return $this->avatar->getThumb($size, $size, $options);
         }
 
-        return '//www.gravatar.com/avatar/' .
-            md5(strtolower(trim($this->email))) .
-            '?s='. $size .
-            '&d='. urlencode($default);
```

Comment:
We can't just get rid of the gravatar entirely, we need to detect when one exists and cache it locally.

Reply:
@LukeTowers can we load it from local instead then? I could save it to file location:

```html
https://github.com/octobercms/october/tree/develop/modules/backend/assets/images
```

Reply:
@ayumihamsaki the data URI isn't the problem, the problem is that people that use gravatars will start seeing mystery man instead of their gravatar. What we should do is add an `afterSave()` method to the user model to attempt to pull in a local copy of the user's gravatar if they don't already have an avatar set. That way existing users just need to go to their profile page and save once to get their gravatar added back.

Reply:
I understand what you mean, I just don't know how to code it.

Reply:
Something like the following:

```php
use October\Rain\Network\Http;

public function afterSave()
{
    if ($this->avatar()->exists()) {
       return;
    }

    $avatarRequest = Http::get('http://www.gravatar.com/avatar/' . md5(strtolower(trim($this->email))) . '?s=' . $size . '&d=404');

    if ($avatarRequest->code === 200) {
         $avatar = new System\Models\File();
         $avatar->fromData($avatarRequest->body, $this->email . '-gravatar.jpg');
         $avatar->save();
         $this->avatar()->attach($avatar);
    }
}
``` 
testing required of course but that should get most of the job done. Btw `http://` is intentional, some servers have broken open SSL installations that cause issues when requesting HTTPS URLs from the server so it's probably better to just request HTTP for this.

Reply:
And if i dont want to download file per every admin? Because server will just have 6 same images if they dont use gravatar

Reply:
> Btw http:// is intentional, some servers have broken open SSL installations that cause issues when requesting HTTPS URLs from the server so it's probably better to just request HTTP for this.

Doesn't work like that, you got Mixed Content API stopping that, then if the certificate is missing or broken the borwser will block it and then lastly if the website doesn't have the redirect in .HTaccess you wouldn't even get the file access.

Again why can't I do this:

https://github.com/octobercms/october/tree/develop/modules/backend/assets/images/avatar-default.jpg

Then you got a stable connection and don't reply on a bunch of what if's for the external connection.

Reply:
@Samuell1 

Is there a piece of code you can link to, which is doing that, or show me in a screenshot, gif, or video. I didn't even know October does that.

Reply:
@LukeTowers e.g.

```php
<?php

public function afterSave()
{
    if ($this->avatar()->exists()) {
       return;
    }

    $avatarRequest =  Url::asset('modules/backend/assets/images/gravatar.jpg');

    if ($avatarRequest->code === 200) {
         $avatar = new System\Models\File();
         $avatar->fromData($avatarRequest->body, $this->email . '-gravatar.jpg');
         $avatar->save();
         $this->avatar()->attach($avatar);
    }
}
```

Reply:
> And if i dont want to download file per every admin? Because server will just have 6 same images if they dont use gravatar. This should as option in admin settings like "Use Gravatar image"

@Samuell1 gravatar will only be used if the user has a gravatar set up for themselves, same as it works currently. 

@ayumihamsaki 

> Doesn't work like that, you got Mixed Content API stopping that, then if the certificate is missing or broken the browser will block both HTTPS and HTTP and then lastly if the website doesn't have the redirect in .HTaccess you wouldn't even get the file access for the HTTP.

It does work like that, the browser is not the one requesting the HTTP URL, as far as the browser is concerned a gravatar image magically appears from the server running October, the browser never talks to Gravatar. Thus, no mixed content warning.

Again why can't I do this:

https://github.com/octobercms/october/tree/develop/modules/backend/assets/images/avatar-default.jpg

Because we're not going to drop support for gravatars out right. Displaying a static image without enabling people to still use their gravatars will break gravatars.

Reply:
@LukeTowers I'm still confused, going to add your code anyway. If @w20k comment gives us the option to turn gravatar support on/off then that's all good.